File: ui/controllers/AnalysisController.cpp
#include "ui/controllers/AnalysisController.h"

#include "core/analyzer/GameAnalyzer.h"
#include "core/scanner/SteamScanner.h"

#include <QtConcurrent/QtConcurrentRun>

namespace gsm::ui {

QFuture<gsm::core::GameAnalysis> AnalysisController::analyzeFolder(const QString& folderPath, const QString& gameName)
{
    const std::string path = folderPath.toStdString();
    const std::string name = gameName.toStdString();

    cancelFlag_->store(false);
    auto flag = cancelFlag_;
    return QtConcurrent::run([path, name, flag]() {
        gsm::core::GameAnalyzer analyzer;
        return analyzer.analyze(path, name, flag.get());
    });
}

QFuture<std::vector<gsm::core::GameEntry>> AnalysisController::scanSteamGames()
{
    return QtConcurrent::run([]() {
        gsm::core::SteamScanner scanner;
        return scanner.scanInstalledGames();
    });
}

void AnalysisController::cancel()
{
    cancelFlag_->store(true);
}

} // namespace gsm::ui

File: game_storage_manager/ui/controllers/analysis_controller.py
from __future__ import annotations

import threading
from concurrent.futures import Future

from game_storage_manager.core.analyzer.game_analysis import GameAnalysis
from game_storage_manager.core.analyzer.game_analyzer import GameAnalyzer
from game_storage_manager.core.scanner.game_entry import GameEntry
from game_storage_manager.core.scanner.steam_scanner import SteamScanner


class AnalysisController:
    def __init__(self):
        self._cancel_flag = threading.Event()

    def analyze_folder(self, folder_path: str, game_name: str = "") -> Future[GameAnalysis]:
        self._cancel_flag.clear()
        flag = self._cancel_flag

        future: Future[GameAnalysis] = Future()

        def _run():
            try:
                analyzer = GameAnalyzer()
                result = analyzer.analyze(folder_path, game_name, flag)
                future.set_result(result)
            except Exception as e:
                future.set_exception(e)

        thread = threading.Thread(target=_run, daemon=True)
        thread.start()
        return future

    def scan_steam_games(self) -> Future[list[GameEntry]]:
        future: Future[list[GameEntry]] = Future()

        def _run():
            try:
                scanner = SteamScanner()
                result = scanner.scan_installed_games()
                future.set_result(result)
            except Exception as e:
                future.set_exception(e)

        thread = threading.Thread(target=_run, daemon=True)
        thread.start()
        return future

    def cancel(self):
        self._cancel_flag.set()
